### PR TITLE
Add hardware crypto module SDF interface (sync code).

### PR DESCRIPTION
**Pull request by MaggieNgWu**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -62,6 +62,10 @@ include(ProjectEVMC)
 include(ProjectVRF)
 include(ProjectEVMONE)
 
+if(SDF)
```
Comment (bxq2011hust) [inline CMakeLists.txt:65]:
```suggestion
if(USE_SDF)
```

File: cmake/Options.cmake [hunk 2]
```diff
@@ -74,13 +74,21 @@ macro(configure_project)
     eth_default_option(TOOL OFF)
     # code coverage
     eth_default_option(COVERAGE OFF)
+    # hardware crypto sdf interface
+    eth_default_option(SDF OFF)
```
Comment (bxq2011hust) [inline cmake/Options.cmake:78]:
```suggestion
    eth_default_option(USE_HSM OFF)
```

File: cmake/Options.cmake [hunk 3]
```diff
@@ -122,6 +130,7 @@ macro(print_config NAME)
     message("-- CMAKE_BUILD_TYPE   Build type                   ${CMAKE_BUILD_TYPE}")
     message("-- TARGET_PLATFORM    Target platform              ${CMAKE_SYSTEM_NAME} ${ARCHITECTURE}")
     message("-- BUILD_STATIC       Build static                 ${BUILD_STATIC}")
+    message("-- SDF                Build SDF HSM                ${SDF}")
```
Comment (bxq2011hust) [inline cmake/Options.cmake:133]:
```suggestion
    message("-- USE_HSM                Build with hardware Secure module                ${SDF}")
```

File: cmake/ProjectSDF.cmake [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+# (c) 2016-2020 fisco-dev contributors.
+#------------------------------------------------------------------------------
+
+include(ExternalProject)
+
+if("${CMAKE_HOST_SYSTEM_NAME}" MATCHES "Linux")
+    if("${ARCHITECTURE}" MATCHES "aarch64")
+        set(SDF_LIB_NAME sdf_crypto_arm.a)
+    else()
+        message(FATAL "unsupported platform")
```
Comment (bxq2011hust) [inline cmake/ProjectSDF.cmake:26]:
```suggestion
        message(FATAL "HSM only support aarch64 Linux, the ${SDF_LIB_NAME} is not supported.")
```

File: cmake/ProjectSDF.cmake [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    endif()
+elseif(APPLE)
+    message(FATAL "unsupported platform")
```
Comment (bxq2011hust) [inline cmake/ProjectSDF.cmake:29]:
```suggestion
    message(FATAL "HSM only support aarch64 Linux, the ${SDF_LIB_NAME} macOS is not supported.")
```

File: cmake/ProjectSDF.cmake [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+else()
+    message(FATAL "unsupported platform")
+endif()
+
+ExternalProject_Add(libsdf
+    PREFIX ${CMAKE_SOURCE_DIR}/deps
+    DOWNLOAD_NAME sdf.tar.gz
+    DOWNLOAD_NO_PROGRESS 1
+    URL https://github.com/MaggieNgWu/sdf-crypto-1/archive/refs/tags/V0.2.7.tar.gz
```
Comment (bxq2011hust) [inline cmake/ProjectSDF.cmake:38]:
```suggestion
    URL https://github.com/FISCO-BCOS/sdf-crypto-1/archive/refs/tags/V0.2.7.tar.gz
```

File: libdevcrypto/sdf/SDFSM2Signature.cpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    if (!_sig->isValid())
+    {
+        return h512{};
+    }
+    if (SDFSM2Verify(_sig->v, _sig, _message))
+    {
+        return _sig->v;
+    }
+    return h512{};
+}
```
Comment (HaoXuan40404) [inline libdevcrypto/sdf/SDFSM2Signature.cpp:172]: add a new libe

File: libdevcrypto/sdf/SDFSM2Signature.cpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    key.setPublicKey((unsigned char*)_keyPair.pub().ref().data(), 64);
+    h256 privk((byte const*)key.PrivateKey(),
+        FixedHash<32>::ConstructFromPointerType::ConstructFromPointer);
+
+    // According to the SM2 standard
+    // step 1 : calculate M' = Za || M
+    // step 2 : e = H(M')
+    // step 3 : signature = Sign(e)
+    // get provider
+    BIGNUM* res = NULL;
```
Comment (cyjseagull) [inline libdevcrypto/sdf/SDFSM2Signature.cpp:51]: Please abstract this option to a single function, and hared with related calls of `SM2`

File: libdevcrypto/sdf/SDFSM2Signature.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    size_t zValueLen = SM3_DIGEST_LENGTH;
+    std::string pubHex = toHex(_pubKey.data(), _pubKey.data() + 64, "04");
+    sm2Group = EC_GROUP_new_by_curve_name(NID_sm2);
+    if ((pubPoint = EC_POINT_new(sm2Group)) == NULL)
+    {
+        throw "[SDFSM2::verify] ERROR of Verify EC_POINT_new";
+    }
+    if (!EC_POINT_hex2point(sm2Group, (const char*)pubHex.c_str(), pubPoint, NULL))
+    {
+        throw "[SDFSM2::verify] ERROR of Verify EC_POINT_hex2point";
```
Comment (cyjseagull) [inline libdevcrypto/sdf/SDFSM2Signature.cpp:117]: this will lead to memory leak, please check here carefully.

File: CMakeLists.txt (hunk 1 above)
Comment (MaggieNgWu) [reply CMakeLists.txt:65]: OK 
USE_HSM_SDF

File: cmake/ProjectSDF.cmake (hunk 4 above)
Comment (MaggieNgWu) [reply cmake/ProjectSDF.cmake:26]: OK，Thanks
、

File: cmake/Options.cmake (hunk 2 above)
Comment (MaggieNgWu) [reply cmake/Options.cmake:78]: OK
USE_HSM_SDF

**Review by bxq2011hust — APPROVED** [review]

**Review by cyjseagull — APPROVED** [review]

**Pull request merged**